### ISSUE TITLE
Serving input function

### ISSUE THREAD

**Issue by sirisha-8** [dmis-lab/biobert#73]
Hi,
Kudos to this Amazing work, I fine-tuned biobert for NER. I got the tensorflow checkpoints.I figured out how to give raw text and get_ner predictions.The only problem here is checkpoint is loading for every input batch.So to avoid this problem ,One of the methods I am aware is to create tensorflow flow saved model(.pb).To create this servering input funtion is required.Using serving function I was able to create model.But I am struck in giving inputs to this model.

`def serving_input_fn():

    label_ids = tf.placeholder(tf.int32, [None], name='label_ids')
    input_ids = tf.placeholder(tf.int32, [None, FLAGS.max_seq_length], name='input_ids')
    input_mask = tf.placeholder(tf.int32, [None, FLAGS.max_seq_length], name='input_mask')
    segment_ids = tf.placeholder(tf.int32, [None, FLAGS.max_seq_length], name='segment_ids')
    input_fn = tf.estimator.export.build_raw_serving_input_receiver_fn({
        'label_ids': label_ids,
        'input_ids': input_ids,
        'input_mask': input_mask,
        'segment_ids': segment_ids,
    })()
    return input_fn`

So, Could you please help me in resolving this issue.It would be of great help. @jhyuklee  @donghyeonk @winkelman @wonjininfo @DSKSD

**Comment by AndreasSaka** [comment]
@sirisha-8 can you share how to do the task "I figured out how to give raw text and get_ner predictions". ?

**Comment by sirisha-8** [comment]
> @sirisha-8 can you share how to do the task "I figured out how to give raw text and get_ner predictions". ?

@AndreasSaka  Here is the snippet
`def input_example(sentences):
        examples= [run_ner.InputExample(guid="", text=tokenization.convert_to_unicode(x), label=tokenization.convert_to_unicode(" ".join(['O' for i in range(0,len(x.split()))]))) for x in sentences] 
        return examples`

    `predict_examples = input_example(pred_sentences)
    predict_file = os.path.join(output_dir, "predict.tf_record")
    tokens = filed_based_convert_examples_to_features(predict_examples,label_list,max_seq_length, tokenizer,predict_file)
                      

    predict_drop_remainder = False
    predict_input_fn = file_based_input_fn_builder(input_file=predict_file,seq_length=max_seq_length,is_training=False,drop_remainder=predict_drop_remainder)
    result = estimator.predict(input_fn=predict_input_fn)`

You have to import run_ner,tokenization .Copy functions mentioned in above snippet from run_ner. 

Thanks

**Comment by jhyuklee** [comment]
Hi, if you are trying to serve BioBERT for NER, check this biomedical NER tool based on BioBERT (https://github.com/dmis-lab/bern). It supports biomedical normalization for multi-type entities, too. You can also plug your new model into the tool as they are based on this BioBERT code. For instance, see below code for how they process inputs:

https://github.com/dmis-lab/bern/blob/dd413b71b8c6a8942cc7036d664bcd11297518d8/server.py#L487-L503

**Comment by sirisha-8** [comment]
Thank you , I will check and let u know

**Comment by RojaBadugu** [comment]
@Sirisha - I have tried the code snippet you shared for raw_text NER prediction but getting issues with estimator 
result = estimator.predict(input_fn=predict_input_fn)
and I tried to define the estimator used in run_ner.py ,then result has got nothing.
Could you please elaborate how you used .Did you integrated with run_ner.py because there are some dependencies with model function and estimator
Thankyou

**Comment by sirisha-8** [comment]
@RojaBadugu 

`run_config = tf.contrib.tpu.RunConfig(cluster=None,master=None,model_dir = output_dir,save_checkpoints_steps=1000,tpu_config=tf.contrib.tpu.TPUConfig(iterations_per_loop=1000, num_shards=8, per_host_input_for_training=tf.contrib.tpu.InputPipelineConfig.PER_HOST_V2))
bert_config = modeling.BertConfig.from_json_file(bert_config_file)
model_fn = model_fn_builder(bert_config=bert_config,num_labels=len(label_list)+1,init_checkpoint=init_checkpoint,learning_rate=5e-5,num_train_steps=None,num_warmup_steps=None,use_tpu=False,use_one_hot_embeddings=False)

estimator = tf.contrib.tpu.TPUEstimator(use_tpu=False, model_fn=model_fn, config=run_config, train_batch_size=32, eval_batch_size=8, predict_batch_size=8)`

**Comment by soumyasv1** [comment]
> > @sirisha-8 can you share how to do the task "I figured out how to give raw text and get_ner predictions". ?
> 
> @AndreasSaka Here is the snippet
> `def input_example(sentences): examples= [run_ner.InputExample(guid="", text=tokenization.convert_to_unicode(x), label=tokenization.convert_to_unicode(" ".join(['O' for i in range(0,len(x.split()))]))) for x in sentences] return examples`
> 
> ```
> `predict_examples = input_example(pred_sentences)
> predict_file = os.path.join(output_dir, "predict.tf_record")
> tokens = filed_based_convert_examples_to_features(predict_examples,label_list,max_seq_length, tokenizer,predict_file)
>                   
> 
> predict_drop_remainder = False
> predict_input_fn = file_based_input_fn_builder(input_file=predict_file,seq_length=max_seq_length,is_training=False,drop_remainder=predict_drop_remainder)
> result = estimator.predict(input_fn=predict_input_fn)`
> ```
> 
> You have to import run_ner,tokenization .Copy functions mentioned in above snippet from run_ner.
> 
> Thanks

what is label_list?what should i define it as?

**Comment by 2ArounD** [comment]
To help people still struggling with making a minimal implemenatation of serving a BioBERT model, I put my version on GitHub:

https://github.com/2ArounD/medical_entity_extraction
